lib/std/time.zig

```diff
@@ -285,3 +201,194 @@ test "Timer" {
     timer.reset();
     try testing.expect(timer.read() < time_1);
 }
+
+pub const Clock = struct {
+    /// Id represents the type of clock source to interact with.
+    /// Each clock sources may start, tick, and change in different ways.
+    pub const Id = enum {
+        /// Returns nanoseconds since Unix Epoch (Jan 1 1970)
```

Comment:
I'd like to add a bit more to the comment here to help folks understand the behavior of the 'realtime' timestamp. I'm thinking:
```
/// Returns nanoseconds since local Unix Epoch (Jan 1 1970).
/// Forward and backward jumps happen non-deterministically when
/// the system time changes (manually, from NTP, leap seconds, or 
/// others). Except for displaying the current time to a user, building any 
/// logic around this timestamp will probably lead to bugs.
/// NOTE: Comparing values across systems is undefined due to 
/// unknown number of Leap Seconds within each timestamp.
/// NOTE: On Windows, the returned value will be offset by the number of
/// Leap Seconds that were declared and then occurred since this std library
/// was compiled into your program.
///
/// TODO: Implement 'Id.atomic' to produce a timestamp which
/// doesn't have the undefined behavior of Leap Seconds. (TAI or GPS Time).
/// 'Id.atomic' would still suffer from NTP or manual time discontinuities.
```
I'd like to try my hand at implementing Id.atomic if folks like the idea.

Reply:
few notes on the doc comment:

> Comparing values across systems is undefined due to 

Would it be better phrased as "unspecified"? "Undefined" implies that if the program uses the cross-system time comparison, then the entire program becomes unsound / could do anything. The former just means the implementation can decide but the program is still sound.

> Implement 'Id.atomic' to produce a timestamp which doesn't have the undefined behavior of Leap Seconds. 

Again, I don't think leap seconds are undefined behavior. For some, it seems to be undesired info. After some minor research, it looks like linux (or POSIX ?) reports the unix timestamp *without* leap seconds. Also TAI seems to be only available on linux and i'm not sure about the availability of GPS for most of the platforms supported by the stdlib currently. If you're up to try and implement it, I would be interested in the results. Just missing context on this all.

Reply:
I'm happy with the word "unspecified".

Regarding leap seconds, I believe that POSIX just doesn't worry about the concept of leap seconds. It pretends that UTC is a timescale that doesn't have leap seconds. This philosophy kinda works, but it's not accurate.
What happens in practice after a leap second is that NTP is used (some time after the leap second occurs) to get an updated time, and the "current time" is updated (a second is added or subtracted). Local Timestamps that occur between when the leap second occured and when the update arrives (usually less than 5min), are off by 1 second. Time deltas measured across the update have discontinuities.

Windows, on the other hand, defines an epoch that explicitly does not have leap seconds (and has no discontinuities when leap seconds occur), but then zig uses a static variable to convert it to the Unix Epoch in UTC. But this (static) conversion will be wrong by 1 second after the next leap second.

So I'll try over the weekend to see if I can put a solution to this into code with Id.atomic.

Reply:
I didn't understand the complexity as much as I thought I did. I spent about 10 hours researching the Linux and Windows implementations of leap seconds and writing up some notes toward a plan, but it's going to take another weekend to have something concrete to share. Fortunately we have some time, as it looks like the next leap second is at least a year away. They might even be eliminated, so I'm planning to defer some complexity of implementation until a confirmed leap second is announced.

Reply:
I spent l ton more time this weekend. So the best way I found to describe the issue is that UTC timestamps from a Unix Epoch are inherently "Fuzzy" in that it's unclear how many leap seconds are in a given timestamp. This is not a problem for date/time strings. We're currently in Utc2018, Windows FILETIME will stay in Utc2018 forever as of 2018 (even after future leap seconds). But timestamps in the past need to be adjusted to be accurate in the current Epoch.
I figured out how to test if a POSIX system supports CLOCK_TAI and build a consistent Utc2018 timestamp from it. Otherwise, we have to assume CLOCK_REALTIME is in Utc2018 until a leap second is declared (which might not happen).

I've created a CalendarTime struct with Gregorian year, month, day, etc.
I've got Clock.id.utc_2018 implemented that exposes a coherent timestamp that ranges from year 1601 - 3800 or so. Timestamps from before the Gregorian calendar are silly. I have conversion functions between these timestamps and the CalendarTime struct.
I have a table of leap seconds and functions to convert between fuzzyUtc and utc_2018.
Next step is to build a function that can initialize a CalendarTime struct from a string to make testing concise. Then I need to write more tests.